### PR TITLE
Program to stress filesystem

### PR DESCRIPTION
**Pull request by hathyaga**
Program to stress filesystem

Running filesystem related tests using LTP test suite

Signed-off-by: Harsha Thyagaraja <harshkid@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by hathyaga** [comment]
[job.txt](https://github.com/avocado-framework-tests/avocado-misc-tests/files/654290/job.txt)

**Comment by PraveenPenguin** [comment]
@hathyaga can this also not a case in https://github.com/avocado-framework-tests/avocado-misc-tests/blob/master/generic/ltp.py (as yaml param )? please consider

**Comment by hathyaga** [comment]
@PraveenPenguin
The existing ltp.py runs a set of default filesystem tests on the current disk. 
I wanted to consider only specific filesystem tests to run on a directory and disk provided by the user.

**Comment by PraveenPenguin** [comment]
@hathyaga yes i know that can we enhance same test case ?

**Comment by hathyaga** [comment]
@PraveenPenguin  let me run the existing ltp.py and check what fs related tests it runs.

**Comment by PraveenPenguin** [comment]
sure i think you need to add some more param in yaml file as releated to fs 

like this ,
    args: '-s fs_di,-s fs_inod'
    mount_point: '/home'
    fs: 'ext4'
in this case you need to have a check if this not possible we can think about custom yaml file

File: io/disk/ltp_fs.py.data/ltp_fs.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+general:
+    disk: '/dev/sda'
+    script: 'runltp'
+    args: '-s fs_di,-s fs_inod'
+    mount_point: '/home'
```
Comment (narasimhan-v) [inline io/disk/ltp_fs.py.data/ltp_fs.yaml]: would be better if mount point can be dynamically created by the script, and not as an input from the user.

File: io/disk/ltp_fs.py.data/ltp_fs.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+    fs: 'ext4'
```
Comment (narasimhan-v) [inline io/disk/ltp_fs.py.data/ltp_fs.yaml]: Consider adding more fs types

File: io/disk/ltp_fs.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        build.make(ltp_dir, extra_args='autotools')
+        ltpbin_dir = os.path.join(ltp_dir, 'bin')
+        os.mkdir(ltpbin_dir)
+        process.system('./configure --prefix=%s' % ltpbin_dir)
+        build.make(ltp_dir)
+        build.make(ltp_dir, extra_args='install')
+
+    def test(self):
+        for val in self.args:
+            self.log.info("creating ext4 filesystem on disk %s" % self.disk)
```
Comment (narasimhan-v) [inline io/disk/ltp_fs.py]: `self.log.info("creating %s filesystem on disk %s" % (self.fs, self.disk)) `

File: io/disk/ltp_fs.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            create_fs = "mkfs.%s %s" % (self.fs, self.disk)
+            self.log.info("%s" % create_fs)
+            ret = process.system(create_fs, shell=True)
+            if ret != 0:
+                self.fail("Created filesystem on %s failed" % self.disk)
+            self.log.info("Mounting disk %s to mount location %s"
+                          % (self.disk, self.mount_point))
+            mounting = "mount %s %s" % (self.disk, self.mount_point)
+            self.log.info("%s" % mounting)
```
Comment (narasimhan-v) [inline io/disk/ltp_fs.py]: not necessary to log the command

File: io/disk/ltp_fs.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            ret = process.system(mounting, shell=True)
+            if ret != 0:
```
Comment (narasimhan-v) [inline io/disk/ltp_fs.py]: you can eliminate `ret` as mentioned before

File: io/disk/ltp_fs.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+            split_lines = (line.split(None, 3)
+                           for line in result.stdout.splitlines())
+            failed_tests = [items[0] for items in split_lines
+                            if len(items) == 4 and items[2] in fail_statuses]
+            if failed_tests:
+                self.fail("LTP tests failed: %s" % ", ".join(failed_tests))
+            elif result.exit_status != 0:
+                self.fail("No test failures detected, but LTP finished with %s"
+                          % (result.exit_status))
+            self.log.info("Removing the filesystem created on %s" % self.disk)
```
Comment (narasimhan-v) [inline io/disk/ltp_fs.py]: Removing filesystem should be handled in `tearDown()`

File: io/disk/ltp_fs.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        process.system('./configure --prefix=%s' % ltpbin_dir)
+        build.make(ltp_dir)
+        build.make(ltp_dir, extra_args='install')
+
+    def test(self):
+        for val in self.args:
+            self.log.info("creating ext4 filesystem on disk %s" % self.disk)
+            create_fs = "mkfs.%s %s" % (self.fs, self.disk)
+            self.log.info("%s" % create_fs)
+            ret = process.system(create_fs, shell=True)
```
Comment (balamuruhans) [inline io/disk/ltp_fs.py]: its good practise to use process.system with try, and with @narasimhan-v's comment mentioned below

File: io/disk/ltp_fs.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            if ret != 0:
+                self.fail("Created filesystem on %s failed" % self.disk)
+            self.log.info("Mounting disk %s to mount location %s"
+                          % (self.disk, self.mount_point))
+            mounting = "mount %s %s" % (self.disk, self.mount_point)
+            self.log.info("%s" % mounting)
+            ret = process.system(mounting, shell=True)
```
Comment (balamuruhans) [inline io/disk/ltp_fs.py]: same here

File: io/disk/ltp_fs.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+            if failed_tests:
+                self.fail("LTP tests failed: %s" % ", ".join(failed_tests))
+            elif result.exit_status != 0:
+                self.fail("No test failures detected, but LTP finished with %s"
+                          % (result.exit_status))
+            self.log.info("Removing the filesystem created on %s" % self.disk)
+            delete_fs = "dd if=/dev/zero bs=512 count=512 of=%s" % self.disk
+            self.log.info("%s" % delete_fs)
+            ret = process.system(delete_fs, shell=True)
+            if ret != 0:
```
Comment (balamuruhans) [inline io/disk/ltp_fs.py]: here

File: io/disk/ltp_fs.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+                self.fail("Unable to delete filesystem on disk %s" % self.disk)
+            self.log.info("Unmounting directory %s" % self.mount_point)
+            unmounting = "umount %s" % self.mount_point
+            self.log.info("%s" % unmounting)
+            ret = process.system(unmounting, shell=True)
+            if ret != 0:
```
Comment (balamuruhans) [inline io/disk/ltp_fs.py]: here

File: io/disk/ltp_fs.py (hunk 6 above)
Comment (hathyaga) [inline io/disk/ltp_fs.py]: If I have a tearDown(), then the script will compile and install ltp for every argument that I pass. So consciously avoided that.

Comment (narasimhan-v) [inline io/disk/ltp_fs.py]: Not an issue. That is how it is supposed to work.
For example, if you decide to multiplex filesystem type, you need to delete each time, which tearDown() will handle.
Also, if there is any error/failure in script, tearDown() will make sure fs is removed(or tried to remove).

File: io/disk/ltp_fs.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        build.make(ltp_dir)
+        build.make(ltp_dir, extra_args='install')
+
+    def test(self):
+        for val in self.args:
+            self.log.info("creating ext4 filesystem on disk %s" % self.disk)
+            create_fs = "mkfs.%s %s" % (self.fs, self.disk)
+            self.log.info("%s" % create_fs)
+            ret = process.system(create_fs, shell=True)
+            if ret != 0:
```
Comment (narasimhan-v) [inline io/disk/ltp_fs.py]: Also, use `ignore_status=True` in `process.run()` and `process.system()`

**Comment by PraveenPenguin** [comment]
@hathyaga me and @narasimhan-v having discussion about this PR and outcome comes that it not good to touch generic ltp stuff as we required some major set up (in this case create file system mounting and un mounting).

so we can have restriction like if there is major change in in setup it self we need to create new Test case for that  and if any other test which is using partially  same setup can extend this test case .
I think that can be ok with you when other test is going to added in this test case :)

One really good point comes as ltp generic as well this test is going to use same mechanism for error or failure catching why not we can have this as a library in avocado   it self .

@santhosh93 any input from you .

waiting v2 for suggested review comment .

Regards
Praveen

**Comment by PraveenPenguin** [comment]
@narasimhan-v are you want to add something ?

**Comment by narasimhan-v** [comment]
@PraveenPenguin I think you have summarized our discussion well.
Next step is checking if an LTP util can be added to avocado.utils. Lets create an issue in avocado for it to check with people for its feasibility and usability.

**Comment by hathyaga** [comment]
@PraveenPenguin @narasimhan-v Sure. I'll send in a V2 for review ASAP

**Comment by hathyaga** [comment]
@PraveenPenguin @narasimhan-v  
Closing this PR. 
New PR: https://github.com/avocado-framework-tests/avocado-misc-tests/pull/223

**Comment by PraveenPenguin** [comment]
i think you forget to press close button :)

**Pull request closed without merging**